cibuildwheel/util.py

```diff
@@ -103,5 +107,39 @@ def get_for_python_version(self, version):
             return self.base_file_path
 
 
+BuildOptions = NamedTuple("BuildOptions", [
```

Comment:
Ew, I didn't know 3.5 only had this style of NamedTuple... 

You know, I've been considering dropping Python 3.5 support on the host side - 3.6 has so many great features (proper pathlib.Path support, f-strings, better typing)...

Reply:
Strongly agree. It may also reduce test time.

Reply:
Are the types important?

Maybe we should keep the types for a separate PR? I have some changes stashed away that are probably completely outdated after all the refactoring, but I could bring try to rebase them. And types are pretty useless without tooling to check them, anyway?

At any rate, these were the types I found
```
def build(project_dir: str, output_dir: str, build_selector: Callable[[str], bool],
          manylinux_images: Dict[str, str], environment: ParsedEnvironment,
          before_build: Optional[str], build_verbosity: int, repair_command: str,
          test_command: Optional[str], test_requires: List[str], test_extras: str) -> None:
```

Reply:
> Are the types important?
>
> Maybe we should keep the types for a separate PR? 

I do not see big difference between current NamedTuple and untyped one.  
Removing python 3.5 should be in separate PR. 

> And types are pretty useless without tooling to check them, anyway (as demonstrated by the tests passing without the correct Optional[str]s

Even vscode have basic type support and provide hinting base on them. Also marking type mismatch is avaliable.

Reply:
> Even vscode have basic type support and provide hinting base on them. Also marking type mismatch is avaliable.

We didn't have that before, either, did we? That's why I'm wondering if that shouldn't be discussed in the context of #258 and a related PR.

Reply:
I'm increasingly open to typing, having had better experiences with it through the VS Code Python extension (you need to *disable* Jedi in their settings, but it's fantastic).

Typing offers three benefits as I see it
  - inline documentation of types (for others reading the source code)
  - metadata for editors to provide auto-complete
  - type checking, to ensure the above is accurate, and to find bugs before the code is run

So for the sake of the first two, I'm happy for this to be a typing.NamedTuple. It's feeling increasingly likely that types will eventually come to the project, so hopefully we'll be able to check them soon, too.

@YannickJadoul you've mentioned in the past that you had played around with project-level types - I'd be really curious to see how that looks if you're able to share the branch?

Reply:
OK, fair enough. I just feels weird to introduce typing in this PR.

@joerick I'll try to push what I had later today, even though it'll be outdated.

Reply:
But this PR introduces types only in new part. Does not type existing one.